src/compiler/checker.ts

```diff
@@ -17822,8 +18113,14 @@ namespace ts {
          *   1. Some constituent is neither a string nor an array.
          *   2. Some constituent is a string and target is less than ES5 (because in ES3 string is not indexable).
          */
-        function checkElementTypeOfArrayOrString(arrayOrStringType: Type, errorNode: Node): Type {
-            Debug.assert(languageVersion < ScriptTarget.ES2015);
+        function getIteratedTypeOfIterableOrElementTypeOfArrayOrString(arrayOrStringType: Type, errorNode: Node, checkAssignability: boolean): Type {
+            const iteratedType = compilerOptions.downlevelIteration && getIteratedTypeOfIterable(arrayOrStringType, /*errorNode*/ undefined);
+            if (iteratedType) {
+                if (checkAssignability && errorNode) {
+                    checkTypeAssignableTo(arrayOrStringType, createIterableType(iteratedType), errorNode);

```

Comment:
This is a check for `--downlevelIteration`. We perform the check for the same reason as mentioned in `checkIteratedTypeOfIterableOrAsyncIterable`:
> even though we have extracted the iteratedType, we will have to validate that the type passed in is actually an Iterable.

This catches a case where our `getIteratedTypeOfIterable` resolution finds an iterated type, but the array or string type is not actually assignable to `Iterable<T>`, reporting an assignability error.